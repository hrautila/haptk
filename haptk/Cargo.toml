[package]
name = "haptk"
version = "0.3.0"
authors = ["Osma S. Rautila"]
edition = "2021"
description = "HAPTK - Haplotype analysis toolkit"
license = "GPL-3.0-only"
readme = "README.md"
homepage = "https://github.com/xosxos/haptk"
repository = "https://github.com/xosxos/haptk"
exclude = [
    "tests/*",
]


[lib]
crate-type = ["cdylib", "rlib"]
path = "src/lib.rs"
doctest = false

[[bin]]
name = "haptk"
path = "src/bin.rs"

[features]
default = ["clap"]
experimental = ["clap", "dep:fishers_exact"]
clap = ["dep:clap", "dep:anstyle"]

[dependencies]
alphanumeric-sort = "1.5"
bgzip = "0.3"
bio = "2.0"
niffler = "2.6"
ndarray = { version = "0.16", features = ["rayon", "blas", "serde"] }
ndarray-npy = { version = "0.9", default-features = false }
termion = "4.0"
color-eyre = "0.6"
rayon = "1.10"
serde = { version = "1.0", features = ["derive"] }
rand = "0.8.5"
svg = "0.13"
csv = "1.3"
itertools = "0.13"
serde_json = "1.0"
tracing = "0.1"
tracing-appender = "0.2"
indexmap = {version = "2.6", features = ["serde"]}
tracing-subscriber = { version = "0.3", default-features = false, features = ["std", "fmt", "time"] }
time = { version = "0.3", default-features = false, features = ["local-offset"] }
petgraph = { version = "0.6", default-features = false, features = ["serde-1"] }
rust-htslib = { version = "0.44", default-features = false, features = ['bzip2'] }
polars = { version = "=0.33.2", default-features = false, features = ["lazy", "is_in", "csv", "fmt", "serde"]}
image = { version = "0.25", default-features = false, features = ['png'] }
statrs = { version = "0.16", default-features = false }
chrono = { version = "0.4.38" }

# Cli
clap = { version = "=4.4.18", optional = true, features = ["derive"] }
anstyle = { version = "1.0.7", optional = true }

# Genome-wide
fishers_exact = { version="1.0.1", optional=true }


[dev-dependencies]
insta = { version = "1.26.0", default-features = false, features = ["yaml"] }



