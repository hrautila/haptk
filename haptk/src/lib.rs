// HATK - Haplotype analysis toolkit
// Copyright (C) 2022  Osma S. Rautila
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

pub mod libs;
pub use libs::{args, error, io, read_vcf, stats, structs, utils};

#[cfg(feature = "clap")]
pub use libs::clap;

/// Graphs implemented in Rust, HST graph will be replaced by the Python implementation
pub mod graphs;

/// HAPTK commands
pub mod subcommands;
