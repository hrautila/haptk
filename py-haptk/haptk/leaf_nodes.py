import itertools

def find_leaf_neighbors(hst):
    # A set of sibling nodes from the tree
    neighbor_sets = []

    for idx in hst.node_indices():
        # A node is leaf if it has no outgoing edges
        if len(hst.out_edges(idx)) == 0:

            # Checks if all siblings have exactly 1 sample
            is_paired_leaf = True

            # Add the first sibling
            neighbors = [idx]

            # Get parents by moving up to adj_direction
            parents = hst.adj_direction(idx, True)

            # Iterate parent for siblings (I think this loops only once in HSTs)
            for parent in parents:
                for sibling in hst.adj_direction(parent, False):
                    # Dont re-add the first sibling
                    if sibling != idx:
                        data = hst.get_node_data(sibling)

                        # Dont add if there is more than 1 sample in the node
                        if len(data['indexes']) == 1:
                            neighbors.append(sibling)
                            pass
                        else:
                            # A sibling has more than 1 sample
                            is_paired_leaf = False

            if is_paired_leaf:
                neighbors.sort()
                neighbor_sets.append(neighbors)

                
    # Sort
    neighbor_sets.sort()

    # Dedup because looping all leaf nodes produces duplicate pairs
    neighbor_sets = list(k for k,_ in itertools.groupby(neighbor_sets))

    return neighbor_sets

def find_average_leaf_node_length(hst):
    leaf_nodes = get_leaf_nodes(hst)

    sum = 0

    for node_idx in leaf_nodes:
        node_data = hst.get_node_data(node_idx)

        # A saturating substraction just to be sure
        length = max(node_data['stop'] - node_data['start'], 0)
        sum += length

    return sum / len(leaf_nodes)

def get_leaf_nodes(hst):
    leaf_nodes = []

    for idx in hst.node_indices():
        if len(hst.out_edges(idx)) == 0:
            leaf_nodes.append(idx)

    return leaf_nodes

   
