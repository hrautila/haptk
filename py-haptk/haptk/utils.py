from ete3 import Tree, TreeStyle, TextFace, NodeStyle
import json
import argparse
import rustworkx as rx
import gzip
import matplotlib as mpl
import numpy as np

from bisect import bisect_left, bisect_right

def find_gt(a, x):
    'Find leftmost value greater than x'
    i = bisect_right(a, x)
    if i != len(a):
        return a[i]
    raise ValueError

def find_lt(a, x):
    'Find rightmost value less than x'
    i = bisect_left(a, x)
    if i:
        return a[i-1]
    raise ValueError

def find_eq(a, x):
    'Locate the leftmost value exactly equal to x'
    i = bisect_left(a, x)
    if i != len(a) and a[i] == x:
        return i
    raise ValueError


def color_fader(c1,c2,mix): #fade (linear interpolate) from color c1 (at mix=0) to c2 (mix=1)
    c1=np.array(mpl.colors.to_rgb(c1))
    c2=np.array(mpl.colors.to_rgb(c2))
    return mpl.colors.to_hex((1-mix)*c1 + mix*c2)

def return_node_style(color, size):
    style = NodeStyle()
    style["fgcolor"] = color
    style["size"] = size
    style["hz_line_type"] = 0
    style["vt_line_type"] = 0
    style["shape"] = "square"
    return style


def hst_to_graph_dict(hst):
    graph_dict = {}
    for node_idx in hst.node_indices():
        neighbors = hst.neighbors(node_idx)
        if not len(neighbors) == 0:
            children_dict = {}

            for nbr_idx in neighbors:
                children_dict[nbr_idx] = hst.get_node_data(nbr_idx)

            graph_dict[node_idx] = children_dict
    return graph_dict

# Originally from https://stackoverflow.com/questions/50003007/how-to-convert-python-dictionary-to-newick-form-format
def newickify(hst, graph_dict, min_samples, hard_cut, min_start, max_stop, right_up) -> str:
    visited_nodes = set()

    # Recursion
    def newick_render_node(node_idx, node_data) -> str:
        assert node_idx not in visited_nodes, "Error: The tree may not be circular!"

        start_stop_clause = False

        if min_start or max_stop:
            # parents = hst.G.predecessors(node_idx)
            # if len(parents) > 0:
                # parent_node = parents[0]
                # start = hst.coords[parent_node['start_idx']]['pos']
                # stop = hst.coords[parent_node['stop_idx']]['pos']
                # print(start, stop, len(parent_node['haplotype']), min_start, max_stop)
            start = node_data['start']['pos']
            stop = node_data['stop']['pos']

            # print(start, stop, len(node_data['haplotype']), min_start, max_stop)

            if min_start and start <= min_start:
                start_stop_clause = True

            if max_stop and stop >= max_stop:
                start_stop_clause = True              

        if node_idx not in graph_dict or len(node_data["indexes"]) < min_samples or start_stop_clause:
            # Leaf node string
            return F'{node_idx}:0'
        else:
            # Add node to visited nodes
            visited_nodes.add(node_idx)

            # Get children
            children = graph_dict[node_idx]

            # Sort keys by size
            keys_and_size = [(child, len(hst.G.get_node_data(child)["indexes"])) for child in children.keys()]
            if right_up:
                keys_and_size.sort(key = lambda x: x[1], reverse=True)
            else:
                keys_and_size.sort(key = lambda x: x[1], reverse=False)

            if hard_cut:
                keys_and_size = list(filter(lambda x: x[1] >= min_samples, keys_and_size))

                if len(keys_and_size) == 0:
                    return F'{node_idx}:0'

            # Children strings recursion
            children_strings = [newick_render_node(child, children[child]) for (child, _) in keys_and_size]
            children_strings = ",".join(children_strings)

            # Parent node string
            return F'({children_strings}){node_idx}:0'

    # Root node is 0
    root_node = 0
    node_data = hst.G.get_node_data(root_node)
    newick_string = newick_render_node(root_node, node_data) + ';'

    # assert visited_nodes == set(graph_dict.keys()), "Error: some nodes aren't in the tree"

    return newick_string

