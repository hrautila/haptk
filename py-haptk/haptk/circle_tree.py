from ete3 import TreeStyle, TextFace, CircleFace,  NodeStyle, add_face_to_node

from haptk import utils

def circle_tree_style():
    ts = TreeStyle()

    ts.branch_vertical_margin = 10
    ts.show_leaf_name = False
    ts.show_scale = False

    ts.mode = "c"
    ts.arc_start = 0 # 0 degrees = 15:00 o'clock
    ts.arc_span = 355
    ts.min_leaf_separation = 0
    ts.root_opening_factor = 0
    ts.optimal_scale_level = "mid"
    ts.allow_face_overlap = True

    return ts

def is_maj_up_to(hst, node, value):
    iter_node = node
    # print(f"checking {iter_node.name} for majority")
    for _ in range(0, value):
        if iter_node.name == "0":
            return True

        if not iter_node.up.name == "0":
            # print("input node:", iter_node.up)
            is_big = is_majority_node(hst, iter_node.up)
            if is_big == False:
                # print(f"run {_}: {iter_node.up.name} is not big")
                return False

        iter_node = iter_node.up
            
    return True

def is_majority_node(hst, node):
    parent = node.up
    input_node_data = hst.get_node_data(int(node.name))
    value = len(input_node_data["indexes"])
    # print(f"input node {node.name} has {value} indexes")

    siblings = {}
    for child in parent.children:
        data = hst.get_node_data(int(child.name))
        if not child.name == node.name:
            siblings[child.name] = len(data["indexes"])


    # print(value, siblings.values())
    for v in siblings.values():
        if value < v: 
            return False
  

def get_children_idx_amounts(hst, node):
    return [len(hst.get_node_data(c.name)["indexes"]) for c in node.children]

def draw_circle_tree(hst, t, output, samples, samples_to_tag, colors):
    style = NodeStyle()
    style["size"] = 0

    narea_color = "#DAF7A6"
    larea_color = "#FFB6C1"

    spent_parents = []
    spent_grand_parents = []

    for n in t.traverse():
        n.set_style(style)
        
        if not n.is_leaf():
            indexes = hst.get_node_indexes(n.name)
            label = len(indexes)
            F = TextFace(label, tight_text=True, penwidth=30, fsize=8)
            F.rotation = 270
            n.add_face(F, column=0)

            count = 0
            for i in get_children_idx_amounts(hst, n):
                if i > 80:
                    count += 1

            if count > 1:
                haplotype = hst.get_node_haplotype(n.name)
                F = TextFace(f"{len(haplotype)}", fsize=30, tight_text=True, fgcolor="red")
                # F.rotation = 270
                n.add_face(F, column=0)
                
                

        if n.is_leaf():
            if samples_to_tag == []:
                n.set_style(utils.return_node_style("#000", 0))
                n.img_style["bgcolor"] = "#FFF"

            for (color_i, sample_list) in enumerate(samples_to_tag):
                results = [i for i in indexes if samples[i] in sample_list]
                if results:
                    n.set_style(utils.return_node_style(colors[color_i], 0))
                    n.img_style["bgcolor"] = colors[color_i]

            if n.up:
                parent = n.up
                parent_indexes = hst.get_node_indexes(parent.name)

                if len(parent_indexes) == 2:
                    if parent in spent_parents:
                        if is_maj_up_to(hst, n, 14):
                            # print(f"{n.name} is_maj")

                            haplotype = hst.get_node_haplotype(n.name)
                            start_idx = hst.get_node_start_idx(n.name)
                            stop_idx = hst.get_node_stop_idx(n.name)
                            start = hst.coords[start_idx]
                            stop = hst.coords[stop_idx]
                            # print(start, stop)
                            bp = stop['pos'] - start['pos']
                            F = TextFace(f"{len(haplotype)}, {bp}", fsize=30, fgcolor="white")
                            # F.margin_left = 50
                            n.add_face(F, column=0)

                            # n.set_style(utils.return_node_style("white", 0))
                            # n.img_style["bgcolor"] = "white"
                    else:
                        spent_parents.append(parent)


    return t

