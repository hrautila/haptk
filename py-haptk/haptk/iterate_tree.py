from ete3 import TreeStyle, TextFace, CircleFace,  NodeStyle, add_face_to_node
from scipy.stats import fisher_exact

from haptk import utils

def iterate_tree_style():
    ts = TreeStyle()

    ts.branch_vertical_margin = 10
    ts.show_leaf_name = False
    ts.show_scale = False

    ts.mode = "c"
    ts.arc_start = 0 # 0 degrees = 15:00 o'clock
    ts.arc_span = 355
    ts.min_leaf_separation = 0
    ts.root_opening_factor = 0
    ts.optimal_scale_level = "mid"
    ts.allow_face_overlap = True

    return ts

def iterate_tree_inner(hst, t, samples_to_tag):
    style = NodeStyle()
    style["size"] = 0

    for n in t.traverse():
        n.set_style(style)
        
        indexes = hst.get_node_indexes(n.name)

        tot_cases, tot_ctrls = 0, 0
        n_cases, n_ctrls = [], []

        for (i, sample_list) in enumerate(samples_to_tag):
            if i == 0:
                tot_cases = len(sample_list)
                n_cases = [i for i in indexes if hst.samples[i] in sample_list]
            if i == 1:
                tot_ctrls = len(sample_list)
                n_ctrls = [i for i in indexes if hst.samples[i] in sample_list]

        n_cases = len(n_cases)
        n_ctrls = len(n_ctrls)
        # res = fisher_exact([[n_cases, n_ctrls], [tot_cases - n_cases, tot_ctrls - n_ctrls]], alternative="greater")
        res = fisher_exact([[n_cases, n_ctrls], [tot_cases - n_cases, tot_ctrls - n_ctrls]])
        
        print(f"{n.name},{res.pvalue},{n_cases},{n_ctrls},{tot_cases},{tot_ctrls}")
        if res.pvalue < 0.0005:
            label = len(indexes)
            F = TextFace(f"{res.pvalue:.4}", tight_text=True, penwidth=30, fsize=8, fgcolor='red')
            F.rotation = 270
            F.background.color = "#FFF"
            F.margin_right = 10
            n.add_face(F, column=0)

        # if n.is_leaf() and res.pvalue < 0.005:
        label = len(indexes)
        F = TextFace(label, tight_text=True, penwidth=30, fsize=8)
        F.rotation = 270

        mix = res.pvalue * 1000
        if mix > 1:
            mix = 1
        color = utils.color_fader('red', 'blue', mix)

        F.background.color = color
        n.add_face(F, column=0)

    return t
