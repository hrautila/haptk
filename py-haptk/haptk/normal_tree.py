from ete3 import TreeStyle, TextFace, NodeStyle

from haptk import utils

def normal_tree_style():
    ts = TreeStyle()
    ts.rotation = 90
    ts.show_leaf_name = False
    ts.show_scale = False
    ts.min_leaf_separation = 0
    ts.branch_vertical_margin = 5
    ts.optimal_scale_level = "full"
    ts.allow_face_overlap = True
    return ts


def draw_normal_tree(hst, t, samples_to_tag, colors, proportions, show_haplotype, n_markers, show_pos):
    style = NodeStyle()
    style["size"] = 0

    for n in t.traverse():
        n.set_style(style)

        node_data = hst.get_node_data(int(n.name))

        label = len(node_data["indexes"])

        if proportions:
            top_data = hst.get_node_data(int('0'))
            samplen = len(top_data["indexes"])
            prc = label / samplen
            label = f"N={label} ({prc:.2f})"
        else:
            label = f"{label}"

        if show_haplotype:
            label = f"N={len(node_data['indexes'])}\n{hst.haplotype_string(node_data, n_markers)}"

        if show_pos:
            label = f"N={len(node_data['indexes'])}"
            haplotype = hst.haplotype_string(node_data, n_markers)
            start = node_data['start']['pos']
            stop = node_data['stop']['pos']
            label = f"{label}\n{haplotype}\nstart: {start}\nstop: {stop}"

        F = TextFace(label, tight_text=True, penwidth=30)
        F.rotation = 270
        n.add_face(F, column=0)

        if n.is_leaf():
            if samples_to_tag == []:
                n.set_style(utils.return_node_style("#000", 1))

            for (color_i, sample_list) in enumerate(samples_to_tag):
                results = [i for i in node_data["indexes"] if hst.get_sample_name(i) in sample_list]
                if results:
                    n.set_style(utils.return_node_style(colors[color_i], 7 - color_i))


    return t

