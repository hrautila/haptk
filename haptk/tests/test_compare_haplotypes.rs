mod common;
use common::{standard_args, OUTDIR};

use std::path::PathBuf;

use haptk::{args::Selection, subcommands::uhst};

#[test]
#[cfg(feature = "clap")]
fn compare_haplotypes() {
    let args = standard_args(Selection::All);
    uhst::run(args, 1, false, None).unwrap();

    let args = standard_args(Selection::OnlyLongest);
    uhst::run(args, 1, false, None).unwrap();

    let files = vec![
        PathBuf::from("tests/results/uhst_mbah.csv"),
        PathBuf::from("tests/results/uhst_shared_core_haplotype_only_longest.csv"),
    ];
    let output = PathBuf::from(OUTDIR);
    let cmd = haptk::clap::SubCommand::CompareHaplotypes {
        haplotypes: files,
        output,
        prefix: None,
        csv: true,
        hide_missing: false,
        tag_rows: false,
        log_and_verbosity: crate::common::silent_verbosity(),
        nucleotides: false,
    };

    haptk::clap::run_cmd(cmd).unwrap();

    let res = std::fs::read_to_string("tests/results/ht_comparison.csv").unwrap();
    insta::assert_yaml_snapshot!(res);
}
