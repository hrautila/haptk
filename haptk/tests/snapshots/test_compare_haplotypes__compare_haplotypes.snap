---
source: haptk/tests/test_compare_haplotypes.rs
expression: res
---
"pos,tests/results/uhst_mbah.csv,tests/results/uhst_shared_core_haplotype_only_longest.csv\nchr9_5_A_C,0,-\nchr9_6_G_T,0,-\nchr9_7_A_C,0,-\nchr9_8_G_T,0,-\nchr9_9_A_C,0,-\nchr9_10_G_T,0,-\nchr9_11_A_C,0,-\nchr9_12_G_T,0,-\nchr9_13_A_C,0,-\nchr9_14_G_T,0,-\nchr9_15_A_C,0,-\nchr9_16_G_T,0,-\nchr9_17_A_C,0,0\nchr9_18_G_T,0,0\nchr9_19_A_C,0,0\nchr9_20_G_T,0,0\nchr9_21_A_C,0,0\nchr9_22_G_T,0,0\nchr9_23_A_C,0,0\nchr9_24_G_T,0,0\nchr9_25_A_C,0,0\nchr9_26_G_T,0,0\nchr9_27_A_C,0,0\nchr9_28_G_T,0,0\nchr9_29_A_C,0,0\nchr9_30_G_T,0,0\nchr9_31_A_C,0,0\nchr9_32_G_T,1,1\nchr9_33_A_C,0,0\nchr9_34_G_T,0,0\nchr9_35_A_C,0,0\nchr9_36_G_T,0,0\nchr9_37_A_C,0,0\nchr9_38_G_T,0,0\nchr9_39_A_C,0,0\nchr9_40_G_T,0,0\nchr9_41_A_C,0,0\nchr9_42_G_T,0,0\nchr9_43_A_C,0,0\nchr9_44_G_T,0,0\nchr9_45_A_C,0,0\nchr9_46_G_T,0,0\nchr9_47_A_C,0,0\nchr9_48_G_T,0,-\nchr9_49_A_C,0,-\nchr9_50_G_T,0,-\nchr9_51_A_C,0,-\nchr9_52_G_T,0,-\nchr9_53_A_C,0,-\nchr9_54_G_T,0,-\nchr9_55_A_C,0,-\nchr9_56_G_T,0,-\nchr9_57_A_C,0,-\nchr9_58_G_T,0,-\nchr9_59_A_C,0,-\n"
